### ISSUE TITLE
Overwriting when Launching Jobs quickly

### ISSUE THREAD

**Issue by atong01** [ashleve/lightning-hydra-template#511]
First off thanks for the great template. I've used it as a structure for 4+ papers at this point. I can safely say that this is an extremely effective tool for conducting ML research. If you are interested in adding things to make this template work even better in a research setting I'd be happy to help.

A somewhat annoying problem I ran into happens when launching jobs too quickly. This is perhaps a niche issue but it was difficult to debug for me so thought I should document it here. 

I'm using this template with SLURM to parallelize experiments. A common bash script for me might be something like this:
```
python src/train.py -m experiment=cfm \
  datamodule=sklearn \
  datamodule.system=scurve \
  seed=42,43,44,45,46 &
python src/train.py -m experiment=cfm \
  datamodule=twodim \
  datamodule.system=moon-8gaussians \
  seed=42,43,44,45,46 &
```

If these jobs happen to launch during the same second then they can end up pointing at the same log directory, somehow both jobs work just fine, run, log, spit out things, but they (can depending on cluster profile) overwrite each other on the filesystem.

This is due to the default log file name which may be equivalent up to seconds for the same `task_name`:
`${paths.log_dir}/${task_name}/multiruns/${now:%Y-%m-%d}_${now:%H-%M-%S}`

There are many potential fixes for this:
* change the log path to `${paths.log_dir}/${task_name}/multiruns/${now:%Y-%m-%d}_${now:%H-%M-%S-%f}` for more precision in the path name
* always make sure you have different task names
* add a small sleep between launching jobs on a cluster. Although I haven't figured out exactly when `hydra` determines `now`, but this fix didn't work consistently for me.

I've gone with 1 + 3 for now.

**Comment by ashleve** [comment]
@atong01 Thanks for documenting the problem with solution!

Great to hear that the template has been effective for your papers :)

> If you are interested in adding things to make this template work even better in a research setting I'd be happy to help.

I would love to hear what problems you usually run into, and if there are some specific areas that you think could be improved for research setting, so feel free to reach out. A dedicated issue or discussion might be a good start.

**Comment by libokj** [comment]
> Check here for my latest (public) uses
> 
> * https://github.com/atong01/conditional-flow-matching
> * https://github.com/atong01/trainable_symmetry
> * https://github.com/lazaratan/dyn-gfn
> 
> Happy to start a dedicated thread, but off the top of my head the most interesting areas of improvement for my type of research, which generally involves many small models and not a few large ones, are:
> 
> * Configuration of Launching via SLURM (or other hydra launcher). For this I suggest adding a `configs/launcher` tree (See [here](https://github.com/atong01/conditional-flow-matching/blob/main/configs/launcher/mila_cluster.yaml) for an example)
> * Easier config of user specific paths. When working on a project with multiple people it's nice to have a single file that configures where everything goes. This is currently `configs/paths/default.yaml` which is tracked. I suggest moving this configuration to an untracked file (I've been using `configs/local/default.yaml`) so that I can tell collaborators to change these paths and they don't end up committing their personal paths by accident. See [here](https://github.com/atong01/conditional-flow-matching/blob/main/configs/paths/default.yaml).
> * Aggregation of runs. This is perhaps the most important part that is missing and somewhat unsolved for me. If I have my act together and all the metrics are logged during training to Wandb then I can easily aggregate and munge things to a table in a notebook (See [here](https://github.com/atong01/trainable_symmetry/blob/main/notebooks/results_eval.ipynb)). What is not solved for me is say I save all my models in a multi run, but then I want to evaluate them on a new metric / dataset / whatever. I currently can't do this efficiently as it requires typing out the config + path for every single model run. I'm currently actively working on how to improve this path.
> * Rerunning models that failed. This is perhaps outside the scope of this template, but say I run 500 jobs in parallel and 5 randomly fail. It's very painful for me to find and rerun those 5.
> 
> Anyway just some thoughts. If you'd be interested in PRs for any of these solved things let me know :)

I have implemented a `CSVMultirunMetricsSummary` to aggregate CSV logger results. Might be useful for you guys. I'm also trying to figure out what the best way is to evaluate multiple models from a multirun on a selection of datasets. Because sweep in hydra does not support top-level dictionary override, it's hard to specify `ckpt_path` and their corresponding `model` in a paired manner (https://github.com/facebookresearch/hydra/issues/1258). Any thoughts? @ashleve @atong01 

```
class CSVMultirunMetricsSummary(Callback):
    """On multirun end, aggregate the results from each job's metrics.csv and save them in metrics_summary.csv."""
    def __init__(self, prefix: str = 'test/'):
        self.prefix = prefix

    # TODO: it doesn't seem to be fault-proof when at least one job execution fails
    def on_multirun_end(self, config: DictConfig, **kwargs: Any) -> None:
        multirun_dir = Path(config.hydra.sweep.dir).resolve()
        csv_metrics_file = Path(config.logger.csv.name) / "metrics.csv"
        ckpt_file = config.callbacks.model_checkpoint.filename
        summary_file_path = multirun_dir / "metrics_summary.csv"

        log.info(f"Aggregating metrics with prefix `{self.prefix}` from subdirectories in {multirun_dir}.")
        # Collect all the override subdirectories that contain metrics.csv from a job's output directory
        override_dirnames = [override_dir.name for override_dir in multirun_dir.glob("*/")]
        # Concatenate them into a single dataframe
        if override_dirnames:
            metrics_summary_df = pd.DataFrame()
            for override_dirname in override_dirnames:
                # Use only columns that start with the specified prefix and 'epoch'
                try:
                    metrics_df = pd.read_csv(multirun_dir / override_dirname / csv_metrics_file,
                                             usecols=lambda col: col.startswith((self.prefix, 'epoch')))
                    # TODO this may be conditional depending on config.job.name
                    metrics_df['epoch'] = metrics_df['epoch'] - 1
                except FileNotFoundError:
                    continue
                # Drop empty rows and then add extra parameter columns
                metrics_df.dropna(inplace=True, subset=[col for col in metrics_df.columns if col != 'epoch'])
                override_dict = dict(item.split('=') for item in override_dirname.split(','))
                metrics_df = metrics_df.assign(**override_dict)
                ckpt_path = (multirun_dir / override_dirname /
                             f"{ckpt_file.format(epoch=int(metrics_df['epoch'].iloc[0]))}.ckpt")
                metrics_df['ckpt_path'] = ckpt_path if ckpt_path.is_file() else None

                metrics_summary_df = pd.concat([metrics_summary_df, metrics_df])

            # Save the aggregated dataframe to metrics_summary.csv
            metrics_summary_df.to_csv(summary_file_path, index=False)
            log.info(f"Saved metrics summary to {summary_file_path}.")
        else:
            log.info(f"No metrics.csv found in subdirectories; skipping metrics summary generation")
```

**Comment by AndyJZhao** [comment]
Hi Alex:
I've not switched to the LightningHydra yet, but I've experienced similar issues when launching multiple runs causing folder conflict. Here's my way of avoiding it by defining a unique ID dubbed **uid**:
In the hydra config file, define working_dir (or any dir you trying to save) whose interpolation is **related to uid**.
```yaml
working_dir: $your/temp/path/${.uid}/ # Note the UID interpolation here.
uid: null # To be generated in the main program
```
Then, in the main Python call, update your **uid** before saving anything using a function like below:
```python
from uuid import uuid4

@hydra.main()
def main(cfg):
    # Generate a Unique ID (UID) to avoid file conflict.
    cfg.uid = cur_time + str(uuid4()).split('-')[0]
    # Other code 
```
Hope this resolves your issue : )
